test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java

```diff
@@ -0,0 +1,431 @@
+                                           for (Map.Entry<InetSocketAddress, byte[]> e : map.entrySet())
+                                               newState.put(toCassandraInetAddressAndPort(e.getKey()), fromBytes(e.getValue()));
+
+                                           Gossiper.runInGossipStageBlocking(() -> {
+                                               Gossiper.instance.applyStateLocally(newState);
+                                           });
+                                           return null;
+                                       }).apply(gossipState);
+        }
+    }
+
+    public static byte[] toBytes(EndpointState epState)
+    {
+        try (DataOutputBuffer out = new DataOutputBuffer(1024))
+        {
+            EndpointState.serializer.serialize(epState, out, MessagingService.current_version);
+            return out.toByteArray();
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static EndpointState fromBytes(byte[] bytes)
+    {
+        try (DataInputBuffer in = new DataInputBuffer(bytes))
+        {
+            return EndpointState.serializer.deserialize(in, MessagingService.current_version);
+        }
+        catch (Throwable t)
+        {
+            throw new RuntimeException(t);
+        }
+    }
+
+    private static class PullSchemaFrom implements InstanceAction
+    {
+        final InetSocketAddress pullFrom;
+
+        public PullSchemaFrom(IInvokableInstance pullFrom)
+        {
+            this.pullFrom = pullFrom.broadcastAddress();;
+        }
+
+        public void accept(IInvokableInstance pullTo)
+        {
+            pullTo.acceptsOnInstance((InetSocketAddress pullFrom) -> {
+                InetAddressAndPort endpoint = toCassandraInetAddressAndPort(pullFrom);
+                EndpointState state = Gossiper.instance.getEndpointStateForEndpoint(endpoint);
+                MigrationManager.scheduleSchemaPull(endpoint, state);
+                MigrationManager.waitUntilReadyForBootstrap();
+            }).accept(pullFrom);
+        }
+    }
+
+    private static class BootstrapAction implements InstanceAction, Serializable
+    {
+        private final boolean joinRing;
+        private final Duration waitForBootstrap;
+        private final Duration waitForSchema;
+
+        public BootstrapAction()
+        {
+            this(true, Duration.ofMinutes(10), Duration.ofSeconds(10));
+        }
+
+        public BootstrapAction(boolean joinRing, Duration waitForBootstrap, Duration waitForSchema)
+        {
+            this.joinRing = joinRing;
+            this.waitForBootstrap = waitForBootstrap;
+            this.waitForSchema = waitForSchema;
+        }
+
+        public void accept(IInvokableInstance instance)
+        {
+            instance.appliesOnInstance((String partitionerString, String tokenString) -> {
```

Comment:
spoke in slack, this is the last remaining comment.  Alex asked for the day to replay to this so we will sync up tomorrow about this.

Reply:
Sure, I think generally it's good to reuse existing code and means. 

Unfortunately, in this case I had to not use `surveyMode` and `auto_bootstrap` as means of passing down the calls since they can only be set once per node (unless we introduce some hacks). We may also want to rethink usage of `cassandra.join_ring` in the future, since it can only be set once per node as well. I'll leave this for some future patch though.

Your suggestion is now implemented here: https://github.com/ifesdjeen/cassandra/commit/137e44e5260648d1cb1a2b81f32269ec22037738